### ISSUE TITLE
Allow customization of which characters to automatically pair

### ISSUE THREAD

**Issue by jugglingnutcase** [atom/bracket-matcher#121]
i'm currently maintaining an Atom language grammar for the PowerShell language, and i received [an issue](https://github.com/jugglingnutcase/language-powershell/issues/37) that complains about automatic quote insertion for the backtick character (<kbd>`</kbd>) in the language. In PowerShell the backtick is used for escapes, so backticks rarely surround a phrase of code, rendering this behavior a small annoyance.

i've started [a discussion in the forum](https://discuss.atom.io/t/can-i-stop-bracket-matcher-from-inserting-double-backticks-from-a-language-grammar/15832/1) about this behavior, and it seems this can be an issue for the LaTeX grammar as well.

Is there a way to turn this behavior off for a particular character without losing automatic quote insertion (is that the term for this behavior?) for the other kinds of quotes? It's a nice feature and losing it for one exception is a bit of a bummer :frowning:

If not, would it be possible to bake this into some sort of language setting? and is this an addition you'd want in this package?
## Relevant links
- the language-powershell that started this issue: jugglingnutcase/language-powershell#37
- [related forum discussion](https://discuss.atom.io/t/can-i-stop-bracket-matcher-from-inserting-double-backticks-from-a-language-grammar/15832/1)

**Comment by requnix** [comment]
This is also an issue in Clojure where `'` is used as an escape character, not a quote, and never needs to be paired.

This causes issues like the following: atom/language-clojure#9

It would be great if we could set this per character pair per grammar. If not, I'd settle for excluding grammars and leave it to someone to create a specialised package that deals exclusively with bracketing and quoting for that grammar a la https://github.com/MarcoPolo/atom-paredit.

**Comment by michaelsproul** [comment]
+1, in Rust a single `'` is used for lifetimes, which only appear in certain contexts:

```
fn something<'a>(&'a self) -> &'a u32 {}
struct Thing<X: 'a> {}
```

Allowing the language packs to adjust the completion rules would be great.

**Comment by spencergibb** [comment]
+1 driving me crazy editing markdown

**Comment by patrickhpan** [comment]
+1 on LaTeX. Driving me crazy because switching between coding (where I am physically dependent on bracket autocompletion) and LaTeX (which is undoable with buggy completion) is really getting annoying because I have to re-enable and disable the package.

**Comment by m-bra** [comment]
+1 for rust

**Comment by derekdreery** [comment]
Does anyone have any ideas about how to achieve this? Could you add the ability to register a boolean function where returning false aborts the bracket add? Something like

```js
bracket_matcher.register_match_function(function(editor, other_params_that_are_useful) {
    if(we_dont_want_a_matching_bracket) {
        return false;
    } else {
        return true;
    }
});
```

PS I'm from Rust

**Comment by derekdreery** [comment]
It should at least be easy to add settings to configure which types of brackets should be matched

**Comment by loveencounterflow** [comment]
+1 for the ability to configure matching character pairs with ease. That should probably include per-language-settings; those would probably optionally have to be made from the respective language-specific plugin. 

Also see https://discuss.atom.io/t/add-bracket-pairs-to-highlighting-auto-insertion/22673; not sure which of the forums is the appropriate one for discussions like this.

**Comment by J-qak** [comment]
+ 1 for Common Lisp, where single `'` is used for "quoting" lists and stuff pretty frequently. I've been trying to disable whole quote-completing feature from my fork of language package with no avail. Anybody know why the following code in my `language-lisp/settings/language-lisp.cson` doesn't work?
`
  "bracket-matcher":
    autocompleteBrackets: false
`

**Comment by slang800** [comment]
@J-qak: if you upgrade Atom that should work, since 26d598d607bdfb899932b45f4b867820a53ba552 has been merged.

**Comment by potto007** [comment]
@michaelsproul, @m-bra, @derekdreery : my frustrations with editing Rust code using Atom is what motivated me to write PR #249. I have some changes to make before it has any hope of acceptance, but wanted to let you know it's out there so you could at least benefit from it and give any feedback you may have.

Given that bracket-matcher is so core to Atom, it stands to reason that it should be highly configurable and as performant as possible.

**Comment by georgd** [comment]
Besides of programming and markup languages the configurability of the bracket matcher is also essential for natural language. Currently it only supports typographical quotation mark matching only for English and French and languages the two oft them share patterns with. For all the other languages there’s no matching at all or faulty matching.

The latter is true for German: There are two sets of quotation marks »« and „“. Using the first one will trigger matching according to French rules at entering the closing mark, resulting in »...«», while using the second set will trigger matching according to English rules at entering the closing mark, resulting in „“” (and the same goes for the single quote variants).

**Comment by potto007** [comment]
@georgd PR #249 should resolve this issue for you. Please observe this animated GIF screencast:

![bracket-matcher-germanquotes](https://cloud.githubusercontent.com/assets/219596/18890981/af8cea60-84c0-11e6-86bf-5c806899487d.gif)

**Comment by 50Wliu** [comment]
Fixed by #249.
